mlkem/poly.c

```diff
@@ -198,17 +198,34 @@ void poly_decompress(poly *r, const uint8_t a[KYBER_POLYCOMPRESSEDBYTES]) {
 #if !defined(MLKEM_USE_NATIVE_POLY_TOBYTES)
 void poly_tobytes(uint8_t r[KYBER_POLYBYTES], const poly *a) {
   POLY_UBOUND(a, KYBER_Q);
-  unsigned int i;
-  uint16_t t0, t1;
 
-  for (i = 0; i < KYBER_N / 2; i++) {
-    t0 = a->coeffs[2 * i];
-    t1 = a->coeffs[2 * i + 1];
-    // REF-CHANGE: Precondition change, we assume unsigned canonical data
-    r[3 * i + 0] = (t0 >> 0);
-    r[3 * i + 1] = (t0 >> 8) | (t1 << 4);
-    r[3 * i + 2] = (t1 >> 4);
-  }
+  for (unsigned int i = 0; i < KYBER_N / 2; i++)
+      // clang-format off
+  ASSIGNS(i, OBJECT_WHOLE(r))
+  INVARIANT(i >= 0)
```

Comment:
Are all those annotations necessary, or can CBMC figure some of this out itself? I'd like to reduce the clutter if possible.

Reply:
Afraid not - CBMC knows that i can be assigned to in the body of the loop, so does not infer its range automatically at the moment. The range constraints on i can be reduced to a conjunction on a single line though, so I will do that.

Reply:
How about introducing an abbreviation like `LOOP(i, lower_bound, upper_bound)` -- still unsatisfactory because it's the same information that's in the `for`, but less text than `ASSIGNS + INVARIANT + INVARIANT`

Reply:
Not sure. I think we can only have ONE ASSIGNS clause, and it has to be "just right" for each loop in particular, including other variables that get changed by the loop body.

Reply:
Can you double-check?

Reply:
Yes.. _exactly_ one ASSIGNS contract must be given.